### PR TITLE
Security #3: Add spire to cluster

### PR DESCRIPTION
**Pull request by lobkovilya**
Signed-off-by: lobkovilya <ilya.lobkov@xored.com>

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [ ] Added integration testing to cover
- [ ] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: security/utils.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
```
Comment (denis-tingajkin) [inline security/utils.go]: 2019

File: security/spire.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
```
Comment (denis-tingajkin) [inline security/spire.go]: save

File: security/manager.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
```
Comment (denis-tingajkin) [inline security/manager.go]: same

File: .mk/spire.mk [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2018 Cisco and/or its affiliates.
```
Comment (denis-tingajkin) [inline .mk/spire.mk]: same

File: security/conf/registration.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (nickolaev) [inline security/conf/registration.json]: Is JSON mandatory? Could we use YAML?

File: test/integration/basic_monitor_crossconnect_test.go [hunk 6]
```diff
@@ -33,11 +33,11 @@ func TestSingleCrossConnect(t *testing.T) {
 	kubetest.DeployNSC(k8s, nodes[0].Node, "nsc-1", defaultTimeout)
 
 	// monitor client for node0
-	eventCh0, closeFunc0 := kubetest.XconProxyMonitor(k8s, nodes[0], "0")
+	eventCh0, closeFunc0 := kubetest.CrossConnectClientAt(k8s, nodes[0].Nsmd)
```
Comment (nickolaev) [inline test/integration/basic_monitor_crossconnect_test.go]: Should these changes be done i a separate PR?

File: security/conf/registration.json (hunk 5 above)
Comment (lobkovilya) [reply security/conf/registration.json]: No, according to documentation `spire-server` doesn't support YAML:
```
Path to a file containing registration data in JSON format (optional).
```

File: k8s/conf/vpn-gateway-nsc.yaml [hunk 7]
```diff
@@ -13,6 +13,7 @@ spec:
         networkservicemesh.io/app: "vpn-gateway-nsc"
         networkservicemesh.io/impl: "secure-intranet-connectivity"
     spec:
+      serviceAccount: nsc-acc
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply k8s/conf/vpn-gateway-nsc.yaml]: If we leave `default` service account for all workloads then we can't distinguish different types of workloads, all they will have same SPIFFE ID. That's not good for further implementation of Authorization

File: test/integration/basic_monitor_crossconnect_test.go (hunk 6 above)
Comment (lobkovilya) [reply test/integration/basic_monitor_crossconnect_test.go]: I introduced `XconProxyMonitor` in #1357 and that changes is just reverting and replacing it with more general solution - `spire-proxy`. I don't think moving `spire-proxy` to separate PR makes sense

File: .cloudtest/aws.yaml [hunk 8]
```diff
@@ -26,4 +26,8 @@ providers:
       stop: |
         make aws-destroy
       prepare: |
+        make spire-start
         make k8s-config helm-init
+        sleep 10
```
Comment (haiodo) [inline .cloudtest/aws.yaml]: sleep 10???

File: docs/env.md [hunk 9]
```diff
@@ -5,6 +5,7 @@
 **NSMD**
 
 * *NSMD_API_ADDRESS* - Specifies IP address and port to start NSMD server (default ":5001")
+* *INSECURE* - Allows to start NSMD in insecure mode (all `grpc.Dial()` will be called with `grpc.WithInsecure()`)
```
Comment (haiodo) [inline docs/env.md]: Please update to include same ENV variable for admission web hook

File: docs/spec/security.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+*coming soon...*
+
+## Implementation details
+Spire consist of two components: 
+* ***spire-agent*** - DaemonSet, has instances on every node, responsible for workload attestation, provides unix socket for certificate obtaining
+* ***spire-server*** - StatefulSet, one per cluster, responsible for *spire-agent* attestation, store information about SPIFFE ID
+
+In order to obtain certificates workload has to mount unix socket
+`/run/spire/sockets/agent.sock`. All volumes will be mounted by AdmissionWebhook in case 
+it discovers annonation `security.networkservicemesh.io: ""`
```
Comment (haiodo) [inline docs/spec/security.md]: It should be always done if NSMgr is in secure mode and admission web hook passed to use security per cluster.

File: docs/spec/security.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+              "value": "sa:workload-service-account"
+            }
+          ],
+          "spiffe_id": "spiffe://test.com/workload",
+          "parent_id": "spiffe://test.com/spire-agent"
+        }
+      ]
+    }
+    ```
+3. Rebuild spire-registration image:
```
Comment (haiodo) [inline docs/spec/security.md]: I think we should not request users to do any kind of rebuild if they want to add new service account mapping, @lobkovilya could you try to play with helm/etc to have a configuration file or files and allow to push/update them to/on spire server.

File: k8s/cmd/admission-webhook/mutate.go [hunk 12]
```diff
@@ -3,33 +3,77 @@ package main
...
+		logrus.Info("No annotations, skip")
+		return okReviewResponse()
 	}
-	if err = checkNsmInitContainerDuplication(metaAndSpec.spec); err != nil {
-		logrus.Error(err)
-		return errorReviewResponse(err)
+
+	var patch []patchOperation
+
+	_, secure := annotations[securityAnnotationKey]
```
Comment (haiodo) [inline k8s/cmd/admission-webhook/mutate.go]: Please use global passed security parameter instead.

File: k8s/conf/debug/nsmgr-debug.yaml [hunk 13]
```diff
@@ -9,9 +9,12 @@ spec:
       app: nsmgr-daemonset
   template:
     metadata:
+      annotations:
+        security.networkservicemesh.io: ""
```
Comment (haiodo) [inline k8s/conf/debug/nsmgr-debug.yaml]: No needed

File: k8s/conf/icmp-responder-nse.yaml [hunk 14]
```diff
@@ -47,3 +53,5 @@ spec:
 metadata:
   name: icmp-responder-nse
   namespace: nsm-system
+  annotations:
```
Comment (haiodo) [inline k8s/conf/icmp-responder-nse.yaml]: No needed

File: k8s/conf/nsc.yaml [hunk 15]
```diff
@@ -20,3 +26,4 @@ metadata:
   namespace: nsm-system
   annotations:
     ns.networkservicemesh.io: icmp-responder?app=icmp
+    security.networkservicemesh.io: ""
```
Comment (haiodo) [inline k8s/conf/nsc.yaml]: No needed

File: k8s/conf/spire/server-configmap.yaml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+      UpstreamCA "disk" {
+        plugin_data {
+          ttl = "12h"
+          key_file_path = "/run/spire/secrets/bootstrap.key"
+          cert_file_path = "/run/spire/config/bootstrap.crt"
+        }
+      }
+    }
+  bootstrap.crt: |
+    -----BEGIN CERTIFICATE-----
```
Comment (haiodo) [inline k8s/conf/spire/server-configmap.yaml]: as discussed this certificates will be removed and generated using helm in next PRs

File: k8s/conf/spire/server-secrets.yaml [hunk 17]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: spire-server
+  namespace: spire
+type: Opaque
+data:
+  bootstrap.key: LS0tLS1CRUdJTiBFQyBQUklWQVRFIEtFWS0tLS0tCk1JR2tBZ0VCQkRBZzJMYnVsWHpRWDFORisyRGkwUkt6TVdmRUdpb0JoaC9mRnB4N3lPRXFrYS8vVHBhZVUzTzUKUUpSWlhkV0hLdWFnQndZRks0RUVBQ0toWkFOaUFBUmFNSDZkSVpMRWhpTE9kdnpqRzdsWVlObVB6U2N2dGJWegpmTi9qeGFITFNacnRqdVlJRXJOOUNTdUFPQzRqaVBSbjdUKzBNZit2eUMwNjBzdXNpbTR6QlllaDdpOXRVRVcxCjdXK1BwZTNwWjRUeVZmQndLOHV6K1p5YTgrcFVyMk09Ci0tLS0tRU5EIEVDIFBSSVZBVEUgS0VZLS0tLS0K
```
Comment (haiodo) [inline k8s/conf/spire/server-secrets.yaml]: as discussed will be updated in next PRs

File: k8s/conf/vppagent-firewall-nse.yaml [hunk 18]
```diff
@@ -44,6 +45,8 @@ spec:
 metadata:
   name: vppagent-firewall-nse
   namespace: nsm-system
+  annotations:
```
Comment (haiodo) [inline k8s/conf/vppagent-firewall-nse.yaml]: no need this

File: k8s/conf/vppagent-icmp-responder-nse.yaml [hunk 19]
```diff
@@ -48,3 +49,5 @@ spec:
 metadata:
   name: vppagent-icmp-responder-nse
   namespace: nsm-system
+  annotations:
```
Comment (haiodo) [inline k8s/conf/vppagent-icmp-responder-nse.yaml]: no need this

File: k8s/conf/vppagent-nsc.yaml [hunk 20]
```diff
@@ -31,3 +32,5 @@ spec:
 metadata:
   name: vppagent-nsc
   namespace: nsm-system
+  annotations:
```
Comment (haiodo) [inline k8s/conf/vppagent-nsc.yaml]: no need this

File: k8s/conf/vppagent-passthrough-nse.yaml [hunk 21]
```diff
@@ -36,6 +37,8 @@ spec:
 metadata:
   name: vppagent-passthrough-nse-1
   namespace: nsm-system
+  annotations:
```
Comment (haiodo) [inline k8s/conf/vppagent-passthrough-nse.yaml]: no need this

File: security/scripts/registration.sh [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+cleanup() {
+    echo "Cleanup entries"
+    rm -rf /run/spire/data/*
+}
+
+/opt/spire/bin/spire-server run -config /run/spire/config/server.conf -registrationUDSPath /tmp/spire-registration.sock &
+
+cleanup
+
+sleep 2
```
Comment (haiodo) [inline security/scripts/registration.sh]: could we have something more smart here?

File: test/kubetest/k8s.go [hunk 23]
```diff
@@ -267,7 +267,9 @@ type K8s struct {
 	apiServerHost      string
 	useIPv6            bool
 	forwardingPlane    string
+	sa                 []string
```
Comment (haiodo) [inline test/kubetest/k8s.go]: Let's choose a better name

File: k8s/cmd/admission-webhook/utils.go [hunk 24]
```diff
@@ -175,3 +182,56 @@ func readRequest(r *http.Request) ([]byte, error) {
...
+	patch = append(patch, patchOperation{
+		Op:    "add",
+		Path:  path,
+		Value: value,
+	})
+
+	return
+}
+
+func addVolume(spec *corev1.PodSpec, added corev1.Volume, basePath string) (patch []patchOperation) {
```
Comment (denis-tingajkin) [inline k8s/cmd/admission-webhook/utils.go]: [minor] consider to apply basePath after creating patches to reduce parameters in methods here

File: k8s/cmd/admission-webhook/utils.go [hunk 25]
```diff
@@ -175,3 +182,56 @@ func readRequest(r *http.Request) ([]byte, error) {
...
+		value = added
+	}
+
+	patch = append(patch, patchOperation{
+		Op:    "add",
+		Path:  path,
+		Value: value,
+	})
+
+	return
```
Comment (denis-tingajkin) [inline k8s/cmd/admission-webhook/utils.go]: [minor] I guess this syntax is not clear. Can we avoid to use non-obvious result parameter?

File: test/applications/cmd/spire-proxy/main.go [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+func newSpireProxy() *spireProxy {
+	workloadAPIClient := workload.NewX509Client(
+		&workload.X509ClientConfig{
+			Addr: &net.UnixAddr{Net: "unix", Name: security.DefaultAgentAddress},
+			Log:  logrus.StandardLogger(),
+		})
+
+	errCh := make(chan error)
```
Comment (denis-tingajkin) [inline test/applications/cmd/spire-proxy/main.go]: why we need this in this method?

File: security/spire.go [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+			}
+			responseCh <- response
+		}
+	}()
+
+	return responseCh
+}
+
+func newResponse(svidResponse *proto.X509SVIDResponse) (*Response, error) {
+	svid := svidResponse.Svids[0]
```
Comment (denis-tingajkin) [inline security/spire.go]: Can Svids be empty?

File: pkg/tools/socket.go [hunk 28]
```diff
@@ -41,8 +45,24 @@ func GetConfig() DialConfig {
 	return cfg
 }
 
+// InitConfig allows init global DialConfig, should be called before any GetConfig(), otherwise do nothing
+func InitConfig(c DialConfig) {
```
Comment (denis-tingajkin) [inline pkg/tools/socket.go]: [minor] will this used in next part of security or can we move it to `integration/utils` because it used only there?

File: security/manager.go [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+	return cm
+}
+
+func (m *certificateManager) exchangeCertificates(obtainer CertificateObtainer) {
+	for {
+		select {
+		case r := <-obtainer.CertificateCh():
+			m.setCertificates(r)
+		case err := <-obtainer.ErrorCh():
+			logrus.Error(err)
```
Comment (denis-tingajkin) [inline security/manager.go]: [minor] Can we use format here to improve detecting that the error was logged here?

File: test/kubetest/pods/test_common.go [hunk 30]
```diff
@@ -7,7 +7,7 @@ import (
 )
 
 // TestCommonPod creates a new alpine-based testing pod
-func TestCommonPod(name string, command []string, node *v1.Node, env map[string]string) *v1.Pod {
+func TestCommonPod(name string, command []string, node *v1.Node, env map[string]string, sa string) *v1.Pod {
```
Comment (denis-tingajkin) [inline test/kubetest/pods/test_common.go]: [minor] Can we use another name instead of sa?

File: security/test/generate_test.go [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"crypto/x509"
+	"testing"
+
+	. "github.com/onsi/gomega"
+)
+
+func TestSimpleCertCreation(t *testing.T) {
+	RegisterTestingT(t)
```
Comment (denis-tingajkin) [inline security/test/generate_test.go]: do not use deprecated API

File: security/test/manager_test.go [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+	"testing"
+	"time"
+
+	"github.com/networkservicemesh/networkservicemesh/security"
+
+	. "github.com/onsi/gomega"
+)
+
+func TestSecurityManager(t *testing.T) {
+	RegisterTestingT(t)
```
Comment (denis-tingajkin) [inline security/test/manager_test.go]: do not use deprecated API

File: security/test/cert_obtainer_test.go [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+)
+
+func TestCertificateObtainer(t *testing.T) {
+	RegisterTestingT(t)
```
Comment (denis-tingajkin) [inline security/test/cert_obtainer_test.go]: do not use deprecated API

**Comment by edwarnicke** [comment]
@lobkovilya How close are we to this being ready to land?

**Comment by edwarnicke** [comment]
> I find the proposed solution very intrusive. It needs changes to yamls and that is nto acceptable. Also we should consider having spire as optional IMO.
> Please consider using the admission webhook to inject the needed sockets and service accounts in case the pod is annotated properly.
> 
> Last but not least - this PR completely lacks any form of documentation PLease ad the proerly through documentation. You are adding a major feature which changes a lot here, lets have it trhough ly documented.

@lobkovilya I am also curious why we are sticking service accounts in the Pod/Deployment specs given the comment [here](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#service-account-tokens) 

> serviceAccountName is usually omitted because this is done automatically.

It appears one usually does this by providing a service account bearer token, not by inserting directly the service account into the Pod spec.

File: dataplane/go.mod [hunk 34]
```diff
@@ -14,13 +13,14 @@ require (
...
 	github.com/opentracing/opentracing-go v1.1.0
 	github.com/rs/xid v1.2.1
-	github.com/satori/go.uuid v1.2.0
 	github.com/sirupsen/logrus v1.4.2
 	github.com/vishvananda/netlink v1.0.0
 	github.com/vishvananda/netns v0.0.0-20190625233234-7109fa855b0f
 	google.golang.org/grpc v1.23.1
 )
 
+replace github.com/census-instrumentation/opencensus-proto v0.1.0-0.20181214143942-ba49f56771b8 => github.com/census-instrumentation/opencensus-proto v0.0.3-0.20181214143942-ba49f56771b8
```
Comment (edwarnicke) [inline dataplane/go.mod]: Why isn't this part of the common replace block?  Why do we have a replace for OpenCencus?

File: dataplane/vppagent/pkg/vppagent/metrics_collector.go [hunk 35]
```diff
@@ -33,7 +34,7 @@ func (m *MetricsCollector) CollectAsync(monitor metrics.MetricsMonitor, endpoint
 }
 
 func (m *MetricsCollector) collect(monitor metrics.MetricsMonitor, endpoint string) {
-	conn, err := tools.DialTCP(endpoint)
+	conn, err := grpc.Dial(endpoint, grpc.WithInsecure())
```
Comment (edwarnicke) [inline dataplane/vppagent/pkg/vppagent/metrics_collector.go]: Have we not secured the Monitor calls yet?

File: dataplane/vppagent/pkg/vppagent/vppagent.go [hunk 36]
```diff
@@ -107,7 +108,7 @@ func (v *VPPAgent) connectOrDisconnect(ctx context.Context, crossConnect *crossc
 	}
 
 	// TODO look at whether keepin a single conn might be better
-	conn, err := tools.DialTCP(v.vppAgentEndpoint)
+	conn, err := grpc.Dial(v.vppAgentEndpoint, grpc.WithInsecure())
```
Comment (edwarnicke) [inline dataplane/vppagent/pkg/vppagent/vppagent.go]: Why the walkback from tools.DialTCP?  And does this break OpenTracing?

File: deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+        plugin_data {
+          {{- if .Values.azure }}
+          kubelet_read_only_port = 10255
+          {{- else }}
+          skip_kubelet_verification = true
+          {{- end }}
+        }
+      }
+    }
+  bootstrap.crt: |
```
Comment (edwarnicke) [inline deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl]: Why are we checking a bootstrap cert into git?

File: go.mod [hunk 38]
```diff
@@ -1,20 +1,8 @@
...
-	github.com/pkg/errors v0.8.1
-	github.com/sirupsen/logrus v1.4.2
-	github.com/uber-go/atomic v1.4.0 // indirect
-	github.com/uber/jaeger-client-go v2.17.0+incompatible
-	google.golang.org/grpc v1.23.1
-)
-
 replace (
+	git.apache.org/thrift.git => github.com/apache/thrift v0.0.0-20180902110319-2566ecd5d999
+	github.com/census-instrumentation/opencensus-proto v0.1.0-0.20181214143942-ba49f56771b8 => github.com/census-instrumentation/opencensus-proto v0.0.3-0.20181214143942-ba49f56771b8
```
Comment (edwarnicke) [inline go.mod]: Why are we depending on OpenCensus and thrift?

File: k8s/cmd/admission-webhook/patches.go [hunk 39]
```diff
@@ -19,54 +19,60 @@ type patchOperation struct {
 }
 
 func createDNSPatch(tuple *podSpecAndMeta, annotationValue string) (patch []patchOperation) {
+	// TODO: now order of containers is important since nsmdp assign proper workspace only to the first container
 	patch = append(patch, addContainer(tuple.spec,
 		[]corev1.Container{
 			{
-				Name:            "nsm-coredns",
-				Image:           fmt.Sprintf("%s/%s:%s", getRepo(), "nsm-coredns", getTag()),
+				Name:            "nsm-dns-monitor",
```
Comment (edwarnicke) [inline k8s/cmd/admission-webhook/patches.go]: Why is the security patch changing the DNS feature?

File: k8s/cmd/nsmdp/nsmdp.go [hunk 40]
```diff
@@ -165,7 +184,8 @@ func startDeviceServer(nsm *nsmClientEndpoints) error {
 		return err
 	}
 
-	grpcServer := tools.NewServer()
+	grpcServer := grpc.NewServer()
```
Comment (edwarnicke) [inline k8s/cmd/nsmdp/nsmdp.go]: Does this break OpenTracing?

File: controlplane/build/Dockerfile.nsmd [hunk 41]
```diff
@@ -9,6 +9,8 @@ ENV GOPROXY=${GOPROXY}
 ENV PACKAGEPATH=github.com/networkservicemesh/networkservicemesh/controlplane
 ENV GO111MODULE=on
 
+RUN apk --no-cache add git
+
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply controlplane/build/Dockerfile.nsmd]: I had strange issue, during resolving dependencies go tried to use git. I will check, probably now it is fine

File: dataplane/go.mod (hunk 34 above)
Comment (lobkovilya) [reply dataplane/go.mod]: According to the style introduces by Artem, we keep replaces to local modules and other replaces in different blocks. 
New go 1.13 fails during build, because OpenCensus has invalid version tag - `v0.1.0`, that's why we have to use replace

File: deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl (hunk 37 above)
Comment (lobkovilya) [reply deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl]: For the first time, to avoid complexity of quite a big PR. There are related comment from Andrey https://github.com/networkservicemesh/networkservicemesh/pull/1380#discussion_r314268283

File: go.mod (hunk 38 above)
Comment (lobkovilya) [reply go.mod]: We are depending on them indirectly through `spire`

File: deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl (hunk 37 above)
Comment (edwarnicke) [reply deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl]: Got it, could we get an issue filed and comment here (and other places with embedded certs) pointing to that issue so if folks come across it in the interim they know we aren't just wildly insecure :)

Comment (lobkovilya) [reply deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl]: For the first time, to avoid complexity of quite a big PR. There are related comment from Andrey https://github.com/networkservicemesh/networkservicemesh/pull/1380#discussion_r314268283

File: go.mod (hunk 38 above)
Comment (edwarnicke) [reply go.mod]: OK... but we haven't changed the dependencies here... we aren't pulling spire into this go.mod file... so I'm still confused...

File: deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl (hunk 37 above)
Comment (lobkovilya) [reply deployments/helm/nsm/charts/spire/templates/agent-configmap.tpl]: > Got it, could we get an issue filed and comment here (and other places with embedded certs) pointing to that issue so if folks come across it in the interim they know we aren't just wildly insecure :)

Sure!

File: go.mod (hunk 38 above)
Comment (lobkovilya) [reply go.mod]: Yes, but we are depending on `pkg` which is depend on `spire`. And the funny thing about `go module` is that during `build` only the last replace matter. All intermediate replaces just ignored

File: k8s/cmd/admission-webhook/patches.go (hunk 39 above)
Comment (lobkovilya) [reply k8s/cmd/admission-webhook/patches.go]: I just swapped containers, as I already mentioned only the first container receives the correct workspace

File: dataplane/vppagent/pkg/vppagent/metrics_collector.go (hunk 35 above)
Comment (lobkovilya) [reply dataplane/vppagent/pkg/vppagent/metrics_collector.go]: This is call to `vpp-agent`, by default it starts `grpc.Server` with insecure mode. Probably we can configure it to work with security. If so I will add it in the next PR

File: k8s/cmd/nsmdp/nsmdp.go (hunk 40 above)
Comment (lobkovilya) [reply k8s/cmd/nsmdp/nsmdp.go]: Yes, it breaks OpenTracing, will fix

File: go.mod (hunk 38 above)
Comment (edwarnicke) [reply go.mod]: Ah... makes sense now.

**Comment by edwarnicke** [comment]
@haiodo @nickolaev You both have 'requested changes' on this PR.  I *think* those are old and now addressed.  Please review and either reiterate your requested changes or release the requested changes hold on the PR if they are already addressed :)

Would be good to finally get this in :)

File: .cloudtest.yaml [hunk 42]
```diff
@@ -47,7 +47,8 @@ executions:
   - name: "Interdomain tests packet-aws"
     env:
       - STORE_POD_LOGS_IN_FILES=true
-      - STORE_POD_LOGS_DIR=/home/circleci/project/.tests/cloud_test/$(cluster-name)
+      - STORE_POD_LOGS_DIR=/home/circleci/project/.tests/cloud_test/$(cluster-name)/logs
```
Comment (haiodo) [inline .cloudtest.yaml]: I suppose we go away from /logs to plain?

**Comment by haiodo** [comment]
@nickolaev @edwarnicke I think we should merge this already :-)

File: .cloudtest.yaml (hunk 42 above)
Comment (denis-tingajkin) [reply .cloudtest.yaml]: yep, we do not use `logs` dir on master branch.

Comment (lobkovilya) [reply .cloudtest.yaml]: Fixed, thank you
